broni/shapes/primitives.py

```diff
@@ -0,0 +1,86 @@
+from . import Shape
+from .. import Trajectory
+
+import numpy as np
+
+
+class Sphere(Shape):
+    def __init__(self, x: float, y: float, z: float, r: float):
+        self.center = np.array((x, y, z))
+        self.radius = r
+
+    def intersect(self, trajectory: Trajectory):
+        dist = np.linalg.norm(self.center - trajectory.data, axis=1)
+        return dist <= self.radius
+
+    def to_mesh(self):
+        u, v = np.mgrid[0:2 * np.pi:20j, 0:np.pi:10j]
+        x = self.center[0] + self.radius * np.cos(u) * np.sin(v)
+        y = self.center[1] + self.radius * np.sin(u) * np.sin(v)
+        z = self.center[2] + self.radius * np.cos(v)
+
+        return np.array((x, y, z), dtype=float)
+
+
+class Cuboid(Shape):
+    def __init__(self, x0: float, y0: float, z0: float, x1: float, y1: float, z1: float):
+        self.p1 = np.array((x0, y0, z0))
+        self.p2 = np.array((x1, y0, z0))
+        self.p3 = np.array((x0, y1, z0))
+        self.p4 = np.array((x0, y0, z1))
+
+        if (x0, y0, z0) == (x1, y1, z1):
+            raise ValueError("p0 is equal to p1, a Cuboid of zero volume is not supported.")
+
+    def intersect(self, trajectory: Trajectory):
+        if len(trajectory.data) == 0:
+            return np.array([])
+
+        u = self.p1 - self.p2
+        v = self.p1 - self.p3
+        w = self.p1 - self.p4
+
+        uP1 = np.dot(u, self.p1)
+        uP2 = np.dot(u, self.p2)
+        if uP1 > uP2:
+            uP1, uP2 = uP2, uP1
+        uO = np.dot(trajectory.data, u)
```

Comment:
This is the same block 3 times, could be a function, @nicolasaunai what name? :)